maven-surefire-common/src/test/java/org/apache/maven/plugin/surefire/AbstractSurefireMojoProvidersInfoTest.java

```diff
@@ -0,0 +1,172 @@
+
+/**
+ * Testing providerInfo applicable behavior.
+ */
+@RunWith( MockitoJUnitRunner.class )
+public class AbstractSurefireMojoProvidersInfoTest
+{
+
+    @Spy
+    private AbstractSurefireMojo mojo;
+
+    @Test
+    public void defaultProviderAreAlwaysAvailable()
+    {
+        ProviderInfo providerInfo = mojo.new JUnit3ProviderInfo();
+        assertThat( providerInfo.isApplicable() ).isTrue();
+    }
+
+    @Test
+    public void dynamicProviderAreAlwaysApplicable()
+    {
+        ProviderInfo providerInfo = mojo.new DynamicProviderInfo( "test" );
+        assertThat( providerInfo.isApplicable() ).isTrue();
+    }
+
+    @Test
+    public void testNgProviderApplicable()
+    {
+        Artifact testNg = mock( Artifact.class );
+        ProviderInfo providerInfo = mojo.new TestNgProviderInfo( testNg );
+
+        assertThat( providerInfo.isApplicable() ).isTrue();
+
+        // no interaction with artifact only reference are checked
+        verifyNoMoreInteractions( testNg );
+    }
+
+    @Test
+    public void testNgProviderNotApplicable()
+    {
+        ProviderInfo providerInfo = mojo.new TestNgProviderInfo( null );
+        assertThat( providerInfo.isApplicable() ).isFalse();
+    }
+
+    //surefire-junit-platform
+
+    @Test
+    public void jUnitPlatformProviderApplicable()
+    {
+        Artifact junitPlatform = mock( Artifact.class );
+        ProviderInfo providerInfo = mojo.new JUnitPlatformProviderInfo( junitPlatform, aTestClassPath() );
+
+        assertThat( providerInfo.isApplicable() ).isTrue();
+
+        // no interaction with artifact only reference are checked
+        verifyNoMoreInteractions( junitPlatform );
+    }
+
+    @Test
+    public void jUnitPlatformProviderNotApplicable()
+    {
+        ProviderInfo providerInfo = mojo.new JUnitPlatformProviderInfo( null, aTestClassPath() );
+        assertThat( providerInfo.isApplicable() ).isFalse();
+    }
+
+    // surefire-junit4
+
+    @Test
+    public void jUnit4ProviderNullArtifacts()
+    {
+        ProviderInfo providerInfo = mojo.new JUnit4ProviderInfo( null, null );
+        assertThat( providerInfo.isApplicable() ).isFalse();
+    }
+
+    @Test
+    public void jUnit4ProviderOnlyJunitDepArtifact()
+    {
+        Artifact junitDep = mock( Artifact.class );
+        ProviderInfo providerInfo = mojo.new JUnit4ProviderInfo( null, junitDep );
+
+        assertThat( providerInfo.isApplicable() ).isTrue();
+
+        // no interaction with artifact only reference are checked
+        verifyNoMoreInteractions( junitDep );
+    }
+
+
+    @Test
+    public void jUnit4ProviderJunit3WithJDepArtifact()
+    {
+        Artifact junit = aArtifact( "3.8.1" );
+        Artifact junitDep = mock( Artifact.class );
+
+        ProviderInfo providerInfo = mojo.new JUnit4ProviderInfo( junit, junitDep );
+
+        // ??? only existing for junitDep in any version is checked
+        assertThat( providerInfo.isApplicable() ).isTrue();
```

Comment:
It looks like the caller's logic guarantees that the artifact coordinates are `junit:junit-dep` and `JUnit4ProviderInfo` makes only the null check. hm...

Reply:
current state - it is done in this way

Reply:
The point of writing the tests is that my eyes stay blind, I should not see the implementation...

Reply:
code assume that input artifacts is correct .. for me looks like mixed responsibility 
`ProviderInfo` should take as input whole classpath of project and take decision on it in one place.

Now we search artifacts in separate methods in `AbstractSurefireMojo` and pass only result to `ProviderInfo` ...
in `AbstractSurefireMojo` should not be any separate logic for it.

But it is another case and place to improvement.